20_classes_lab/03.py

```diff
@@ -1,13 +1,72 @@
-"""
-Following code assumes a class named Widget that depends on
-other widgets
-Widget ctor takes a name, and function 
-add_dependency(depends)
-marks widget as depends on all widgets in "depends" list
-
-When building a widget, you need to first build all the widgets it
-depends on.
-
-Implement widget so the following code works
-"""
-
+#Widget
+
+class Widget(object):
+    
+
+
+    #Constructor
+    def __init__(self, name):
+        self._name = name
+        self._dependDict = {}
+        self._isBuilt = False
+        self._index = 0   #Index to keep on the inserting order
+
+
+    #Check if Widget object is in the dependencies list
+    def _isDepend(self, name):
+        isDepend = False
+
+        for _, val in self._dependDict.items():
+            if val._name == name or val._isDepend(name):
+                isDepend = True
+                break
+        return isDepend
+
+    #Check circular dependencies
+    def _isCircular(self, widgetObj):
+        return widgetObj._isDepend(self._name)
+
+    #Add dependencies
+    def add_dependency(self, *widgetObjLis):
+        for obj in widgetObjLis:
+            if not self._isCircular(obj):   #Add only if there is no circular dependency
+                self._dependDict[self._index] = obj
+                self._index += 1
+            else:
+                raise Exception("Circular dependencies %s depends on %s" % (obj._name, self._name))
+
+
+    #Build all widgets
+    def build(self):
+        self._isBuilt = True
+        for _, val in self._dependDict.items():
+            if not val._isBuilt: #Build only objects that haven't been built yet
+                val.build()
+                print val._name
+
+
+
+luke    = Widget("Luke")
+hansolo = Widget("Han Solo")
+leia    = Widget("Leia")
+yoda    = Widget("Yoda")
+padme   = Widget("Padme Amidala")
+anakin  = Widget("Anakin Skywalker")
+obi     = Widget("Obi-Wan")
+darth   = Widget("Darth Vader")
+_all    = Widget("All")
+
+
+luke.add_dependency(hansolo, leia, yoda)
+leia.add_dependency(padme, anakin)
+obi.add_dependency(yoda)
+darth.add_dependency(anakin)
+
+_all.add_dependency(luke, hansolo, leia, yoda, padme, anakin, obi, darth)
+_all.build()
+     
+
+
+
+  
+

```

Comment:
👍 this looks excellent. 

A few comments:
* _index seems to be unnecessary. You could get the next number by calling `len(self._dependDict)` but:
* If you care about the order then it is usually better to have a list and not a dictionary